### PR TITLE
Fix inplace op circle reference bug

### PR DESCRIPTION
**Pull request by wyg1997**
解决  https://github.com/Oneflow-Inc/oneflow/pull/5551#issuecomment-898860540 提出的问题。

例子中：

```python
import oneflow as flow
for i in range(1000):
    a = flow.ones((256*256, 1024), requires_grad=True).to("cuda")
    b = flow.ones((1024), requires_grad=True).to("cuda")
    a += b
```

由于a是requires_grad=True又调用了inplace op，后向图中形成了循环引用 `a -> FunctionNode -> ctx -> a`。

在后向图构建过程中只用捕获 tensor.detach() 参与后向计算就可以了，同时保留requires_grad以保证高阶导正确。后面就可以正常做BroadcastAdd了。

### CODE REVIEW

File: oneflow/api/python/functional/python_functions.cpp [hunk 1]
```diff
@@ -85,11 +85,26 @@ py::object PyAdd(py::args py_args, py::kwargs py_kwargs) {
...
+              int dim_b = b_shape->At(index);
+              CHECK_OR_RETURN(dim_a == dim_b || (dim_a > 0 && dim_b == 1))
+                  << "Invalid expand shape " << a_shape->ToString();
+            } else {
+              CHECK_GT_OR_RETURN(a_shape->At(i), 0)
+                  << "Invalid expand shape " << a_shape->ToString();
+            }
+          }
         }
-        return functional::BroadcastAdd(a, b);
```
Comment (hjchen2) [inline oneflow/api/python/functional/python_functions.cpp:92]: 把这段逻辑放到BroadcastAdd functor里去

File: python/oneflow/nn/modules/linear.py [hunk 2]
```diff
@@ -111,11 +108,8 @@ def reset_parameters(self) -> None:
 
     def forward(self, x):
         res = flow.F.matmul(x, self.weight, transpose_a=False, transpose_b=True)
-        if self.use_bias:
-            # TODO(zwx): Inplace add is not supported yet with consistent tensor,
-            # use non-inplace version before inplace broadcast add has been implemented.
-            # res += self.bias
-            res = res + self.bias
```
Comment (wyg1997) [inline python/oneflow/nn/modules/linear.py:118]: @leaves-zwx 这里BroadcastAdd支持inplace了，直接注掉用inplace就可以了吧？可以直接用BroadcastAdd的SBP推导，应该不会出现S->B的情况了

File: oneflow/core/framework/op_expr_grad_function.h [hunk 3]
```diff
@@ -56,11 +56,19 @@ class OpExprGradFunction : public OpExprGradFunctionIf {
                         const OpExprInterpContext& interp_ctx) const override {
     StateT* state = dynamic_cast<StateT*>(ctx);
     CHECK_NOTNULL_OR_RETURN(state);
+    // Only catures detach tensor for calculating grad and set right requires_grad
```
Comment (hjchen2) [inline oneflow/core/framework/op_expr_grad_function.h:59]: catures -> capture

Comment (hjchen2) [reply oneflow/core/framework/op_expr_grad_function.h:59]: detach -> detached

File: oneflow/api/python/functional/python_functions.cpp (hunk 1 above)
Comment (wyg1997) [reply oneflow/api/python/functional/python_functions.cpp]: done

File: oneflow/core/framework/op_expr_grad_function.h (hunk 3 above)
Comment (wyg1997) [reply oneflow/core/framework/op_expr_grad_function.h]: done

**Review by hjchen2 — APPROVED** [review]
